File: list.cpp

#include <iostream>
#include <string.h>
using namespace::std;
class node 
{
private:
//int data;

public:
node* Left;
node* Right;
int data;

node() { data=0;Left=NULL;Right=NULL;}
void setdata(int d)
{ data=d;}

int getdata()
{return data; }

};


class list
{
public:
	node* head, * tail;
	
public:
	list()
	{
		head = NULL; tail = NULL;
	}
	
	void addnode(int data)
	{
		node* n = new node();
		n->setdata(data);
		if (head == NULL)
		{ head = n; tail = n;}
		else
		{
			n->Left = tail;
			tail->Right = n;
			tail=n;			
		}
			
	}
	
	void displaynode()
	{
		node* n = head;
		while(n!=NULL)
		{
			cout << n->getdata() << " ";
			n=n->Right;
		}
	}
	
	void reverse()
	{
		
		node* n1 = head;
		node* n2 = n1;
		node* temp1, *temp2;
				while(n1)
				{
					temp1=n1->Right;
					temp2=n1->Left;
					n1->Left=temp1;
					n1->Right=temp2;
					n2=n1;
					n1 = temp1;					
				}
				head = n2;
	}
	
	void revlist()
	{
		node* n = head;
		node* prev = NULL;
		node* temp = NULL;
		
		while(n)
		{
			temp = n->Right;
			n->Left = temp;
			n->Right = prev;
			prev=n;
			n=temp;			
		}
		head = prev;
	}
};

node* getmiddle(node* h)
{
	if(!h)
		return h;
	node* f = h;
	node* s = h;
	
	while((f->Right!=NULL) && (f->Right->Right != NULL))
	{
		s = s->Right;
		f=f->Right->Right;
	}
	return s;
}

node* merge(node* a, node* b)
{
	node* d = NULL;
	node* r;
	node* p = NULL;

	
	while((a!=NULL) && ( b!=NULL))
	{
		node* t = new node();

        if(a->data > b->data)
        {
        	t=b;
        	b=b->Right;
        }
        else
        {
        	t=a;
        	a=a->Right;
        }
        if(!d)
        {
        	d=t;
        	r=d;
        	p=d;
        	
        }
        else
        {
        	p=d;
        	d->Right = t;
        	d=d->Right;
        	d->Left = p;
        }
	}
	while(b !=NULL)
	{
		node* t = new node();
		t=b;
		p=d;
    	d->Right = t;
    	d=d->Right;
    	d->Left = p;
    	b=b->Right;		
	}
	
	while(a != NULL)
	{
		node* t = new node();
		t=a;
		p=d;
    	d->Right = t;
    	d=d->Right;
    	d->Left = p;
    	a=a->Right;		
		
	}
	
	return r;
	
}


node* mergesort(node* h)
{
	if ( (h==NULL) || (h->Right == NULL))
	{
		return h;
	}
	node* m = getmiddle(h);
	node* half = m->Right;
	m->Right = NULL;
	
	return (merge(mergesort(h), mergesort(half)));
	
}

int main()
{
	int i=0, j=0;
	list l;
	
	cout << "Enter data \n";
	while(i<6)
	{
		cin >> j;
		l.addnode(j);
		i++;
	}
	l.displaynode();
	
	//l.revlist();
	//cout << " \n";
	//l.displaynode();
	
	node* n = mergesort(l.head);
	cout << "\n\n";
	
		while(n!=NULL)
		{
			cout << n->getdata() << " ";
			n=n->Right;
		}
}

File: List.java
package linked;

import java.io.*;
import java.util.*;
import java.lang.String;
import java.util.Scanner;



class List
{
	private Node head;

	public List()
	{
		head = null;
	}	

	public void addnode(int d)
	{
		Node n = new Node(d);
		if(head != null)
		{
			Node end = head;
			while(end.next != null)
			{
				end = end.next;
			}
			end.next = n;
		}
		else
		{
			head=n;
		}
	}

	public void reverse()
	{
		Node start = head;
		Node temp = null;
		Node fwd = head;
		
		while(start != null)
		{
			fwd=start.next;
			start.next = temp;
			temp=start;
			start=fwd;
		}
		head = temp;
	}


	public void display()
	{
		Node start = head;
		while(start != null)
		{
			System.out.println(start.data);
			start=start.next;
		}
	}
	


	public static void main(String[] args)
	{
		System.out.println("List Starts");

		List l = new List();
		l.addnode(10);
		l.addnode(20);
		l.addnode(30);
		l.addnode(40);

		l.reverse();

		l.display();

	}

	class Node
	{
		private int data;
		private Node next;

		public Node()
		{
			data=0;
			next=null;
		}
		public Node(int d)
		{
			data = d;
		}
	}
}